File: octo/oswatchRefactor/src/main/scala/com/sankuai/octo/oswatch/service/LogCollectorService.scala
package com.sankuai.octo.oswatch.service


import com.sankuai.octo.oswatch.utils.Common
import com.sankuai.octo.oswatch.model.LogCollectorResult
import org.slf4j.{LoggerFactory, Logger}
import play.api.libs.json._
//import dispatch._, Defaults._
import scalaj.http._
import scala.concurrent.{duration, Await}
import scala.concurrent.duration.Duration
import scala.collection.mutable.Map

/**
 * Created by dreamblossom on 15/9/30.
 */
object LogCollectorService {

  def getCurrentQPS(appkey: String, method: String, env: String, timestamp: Long, periodInSeconds: Int) = {
    val rv = Map[String, Double]()
    val spanName = if (method == null) "all" else method
    val end = timestamp / Common.ONE_SECOND_IN_MS
    val start = end - periodInSeconds
    val result = getHttpContent(mergePerfQuery(appkey, spanName, env, start.toString, end.toString))
    result.consumer2QpsList.foreach{ consumer =>
      rv += (consumer.consumerAppKey -> consumer.qpsAvg)
    }
    rv
  }

  def mergePerfQuery(appkey: String, spanName: String, env: String, start: String, end: String) = {
    "/api/query?provider=" + appkey +
      "&spanname=" + spanName +
      "&env=" + env +
      "&start=" + start +
      "&end=" + end
  }

  def getHttpContentList(URL: String) = {
    // val feature = Http(url(Common.logURL + URL) > as.String)
    // Json.parse(Await.result(feature, timeout)).as[List[LogCollectorResult]]
    println("request: " + Common.logURL + URL)
    val response: HttpResponse[String] = Http(Common.logURL + URL).option(HttpOptions.readTimeout(Common.httpTimeoutInterval * Common.ONE_SECOND_IN_MS))
      .asString
    println("response: " + response)
    (Json.parse(response.body)).as[List[LogCollectorResult]]
  }

  def getHttpContent(URL: String) = {
    println("request: " + Common.logURL + URL)
    val response: HttpResponse[String] = Http(Common.logURL + URL).option(HttpOptions.readTimeout(Common.httpTimeoutInterval*Common.ONE_SECOND_IN_MS))
      .asString
    println("response: " + response)
    //val feature = Http(url(Common.logURL + URL) > as.String)
    (Json.parse(response.body)).as[LogCollectorResult]
  }
}

File: octo/oswatch/src/main/java/com/sankuai/octo/oswatch/service/LogCollectorService.java
package com.sankuai.octo.oswatch.service;

import com.sankuai.octo.oswatch.model.ConsumerResult;
import com.sankuai.octo.oswatch.model.LogCollectorResult;
import com.sankuai.octo.oswatch.model.PerfQPSResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by chenxi on 6/30/15.
 */

public class LogCollectorService extends QPSReader {
    public Logger logger = LoggerFactory.getLogger(LogCollectorService.class);

    public LogCollectorService(String queryURL) {
        this.queryURL = queryURL;
    }

    public Map<String, Double> getCurrentQPS(String appkey, String method, String env, long timestamp, int periodInSeconds) {
        long start, end;
        String spanname;
        Map<String, Double> rv = new HashMap<String, Double>();

        if(method == null) spanname = "all";
        else spanname = method;

        end = timestamp / ONE_SECOND_IN_MS;
        start = end - periodInSeconds;

        String request = mergePerfQuery(appkey, spanname, env, String.valueOf(start), String.valueOf(end));
        logger.info("request",request);
        LogCollectorResult qpsObj = getHttpContent(request, LogCollectorResult.class);

        if (qpsObj == null) return rv;

        for (ConsumerResult cResult: qpsObj.getConsumer2QpsList()){
            logger.info("consumer:"+cResult.getConsumerAppKey());
            logger.info("qps:"+ Double.valueOf(cResult.getQpsAvg()));
            rv.put(cResult.getConsumerAppKey(), Double.valueOf(cResult.getQpsAvg()));
        }

        return rv;
    }

    public String mergePerfQuery(String appkey, String spanname, String env, String start, String end) {
        return "/api/query?provider=" +
                appkey +
                "&spanname=" +
                spanname +
                "&env=" +
                env +
                "&start=" +
                start +
                "&end=" +
                end;
    }

//    public static void main(String[] args) {
//        LogCollectorService s = new LogCollectorService("http://192.168.3.163:8930");
//        long x = System.currentTimeMillis();
//        while (true) {
//            Map<String, Double> rv = s.getCurrentQPS("com.sankuai.chenxi.test_provider_a", "HelloInterface.hi", x , 5);
//            try{Thread.sleep(1000);} catch (Exception e) {}
//            System.out.println(rv);
//        }
//    }
}
